### ISSUE TITLE
Moving to akka-http

### ISSUE THREAD

**Issue by hntd187** [spark-jobserver/spark-jobserver#863]
Okay friends, I'm gonna spearhead the effort to update our web services from Spray to something that isn't spray.

## Background
For those unaware, Spray has not been maintained for some time because the creator of Spray joined Typesafe/Lightbend in order to work on akka-http, which he now maintains. As a result, we are using an un-maintained web-service framework, which while spray works great, if any issues come up in the future we basically are on our own. So in order for us to move forward we are going to have to refactor these entire parts in order to use another framework.

## Basic Goals
So while I am working on this I have some goals I want to hit along the way.
- [ ] Completely remove our dependency on Spray (including spray-json)
- [ ] Move us to another web framework. For now that will be akka-http unless someone presents a compelling case for another framework, which I'm more than open to listening to as I'm not even sure akka-http is the best. Also same with a json serialization library, something fast that is actively maintained is the minimum.
- [ ] Remove the akka-app project
    - [ ] As a result of removing this I will have to rework and update the metrics we have, but I think that will be easier than it seems. I believe the http://kamon.io guys agreed to help us if we used their library, which sounds great to me.
- [ ] Hopefully get some general refactoring and pain-point refactoring done with the web api (PLEASE TELL ME WHAT YOU'D LIKE HERE IF YOU HAVE A SUGGESTION)

One thing I am going to hard line on this issue though is no new functionality. This is not meant to add new functionality, this is meant only to get us to a maintained web framework, so please no suggestions for functionality additions/deletions, unless there is a super good reason, I will not even consider them. Bug fixes on the other hand I can try and squash them as much as I can along as the way as long as I know about them.

So, I think that is all I have for now, I'll be fleshing out my TODO: list more as I dig into this, but I will most likely be working on a branch here, which will encompass the effort. Mind you, this branch should be considered ultra experimental, not supported and very very not for production or any real work. This is gonna take a lot of time and testing.

**Comment by velvia** [comment]
Thanks @htnd187….  this will be a much-needed upgrade.  Along with that, update to Akka 2.4.x at least.

**Comment by hntd187** [comment]
@velvia the thing is though, akka 2.4.x is java8 and scala 2.11+ only so there is that too...

**Comment by velvia** [comment]
That is true but I believe akka-http requires at least akka 2.4.x.

Another good choice I have used before is scalatra.  It supports different versions of akka, but its DSL is completely different from spray.

There are no shortage of good JSON libraries of course.  This might be a good time to wonder how to let users provide serializations of their own classes.

**Comment by amarouni** [comment]
Hello guys,
I've already asked the question before #711 but since @hntd187 proposed a wider change I'll close mine and move the discussion here.

I'm actually in favor of replacing Spray with a new framework like akka-http (when I played with it was Ok) and to take the opportunity to do some refactoring specially on the `webAPI.scala` : 

- RESTfull responses 
- Proper serialization (no more Maps -> String) 
- Separation (splitting the API)
- Swagger (or other)
- ...

What you think about using the Play framework, even if it's somehow an overkill for SJS ?

I'd be glad to and help on this one.

**Comment by hntd187** [comment]
@velvia perhaps would could experiment with users providing their own serialization? Perhaps some lower level API that allows us to serialize return types or something

@amarouni No, play is way too much. We wouldn't even use 95% of the framework and while I have no opinion of play as a framework I don't think it's the right choice here. Can you elaborate a bit more on splitting the api and what you mean?

**Comment by velvia** [comment]
@amarouni maybe you can help @htnd187 once we have something like a PR or preview somewhere?  Although that might end up being a follow on change.

The biggest advantage of akka-http is compatibility with spray, I would expect the least amount of changes compared to another framework.  Being able to integrate closely with actors is another advantage.  I believe akka-http is based on akka-streams and I really believe streams and reactive is the future.  Some of the things currently done with actors could probably move over to a more reactive API, such as status updates.

**Comment by amarouni** [comment]
@hntd187 Yeah Play is an overkill, although it provides its own JSON serialization that's quiet good. In fact akka-http is the perfect fit, I played with it recently and it's quiet easy and light weight. For the serialization I used Spray-json but since the projects is not maintained anymore it might not be the best solution.

What I meant by splitting the API is simply refactoring the huge `webAPI.scala` into more manageable pieces with all the required refactoring required.

**Comment by velvia** [comment]
Yes, agreed WebApi should be refactored, and perhaps that could be part of @htnd187’s and this common change.

**Comment by hntd187** [comment]
I'm probably gonna have to rewrite most of webapi.scala so yea we can do that refactoring.

**Comment by amarouni** [comment]
@hntd187 @velvia and add proper serialization with separate models and the JSON protocols (if we stick with spary-json) in separate packages.

**Comment by velvia** [comment]
+1.   

We might want to move on from spray-json.  There are tons of good json libs.

**Comment by hntd187** [comment]
I started writing it with spray because akka-http comes with a lot of conveniences for it, but I'm not far enough along for it to matter yet, if someone has a suggestion for another json lib I'm not particular about spray.

**Comment by hntd187** [comment]
Well me looking around a little bit spray comes out of the gate with support in akka-http, and unless I missed something it would seem any other json library would require us to write the interface to akka-http ourselves, which in of itself is not a big deal, but just something to keep in mind as we go forward. We don't deal with gigantic json inputs or outputs so it's conceivable that any library will be "fast enough"

**Comment by amarouni** [comment]
@hntd187 Yeah That's what I noticed also, even in the official akka-http documentation they use spary-json  in all of their examples. If we can get all of the models + spary-json serialization logic into a separate package then it would be easier to plug a new json library if we don't want to stick with it.

**Comment by hntd187** [comment]
@amarouni it's almost as if the akka-http committers have some relationship with spray 🤔 Either way, Ive already got some skeleton code like that going. It's just a bit rough to make the tests 1:1 since we built custom maps before so our JSON didn't fit back to the case classes perfectly but it's not a big deal, I'm trying not to break the current actor contracts at all.

**Comment by hntd187** [comment]
The branch can be found now at https://github.com/spark-jobserver/spark-jobserver/tree/akka-http-expirimental this is where I'll be checking things into.

**Comment by marcoboi** [comment]
Hi @htnd187.
Thanks for putting so much effort into this.
I have identified a few points that would benefit from some work.
It might be a good idea to tackle them at this stage since you're dealing with a substantial refactoring.
I'm happy to discuss them if you like and contribute if needed.

* HTTP responses are not consistent across APIs. For example:
** DELETE for a job will return: { "status": "KILLED" }
** DELETE for data will return "OK"
** DELETE for context will return { "status": "SUCCESS", "result": "Context stopped" }
These and similar inconsistencies makes it difficult to parse the response on the Client side.
This is because the same action results in very different response patterns depending on the API or even on the status of the server (e.g., GET on a started vs completed vs failed job will return responses that differ over several fields).
I believe creating classes for the HTTP responses would greatly improve consistency and streamline future development.
It will also make SJS more useable and more easy to integrate with other applications.
* APIs, DELETE should not take a URL escaped resource id: currently, performing a DELETE on the "data" endpoint requires passing a URL escaped string to point to the resource name. Since the resource ID name contains "slash" signs, most reverse-proxy servers will not let that request through. As a simple fix, I propose identifying resources using a UUID generated by the server and passed to the client in the HTTP response to the POST request. That way the client won't have to pass an ID during POST and won't need any escaping when sending a DELETE.
* HTTP override for DELETE methods: most proxy servers and some firewalls won't let DELETE requests through. Most modern application provide HTTP override to circumvent this limitation. I will provide code to do the override for Spray in the main branch, but it probably make sense to plan this functionality here too as it would only be a few lines addition.
* Splitting the data, jobs, context and binary in different classes is a much needed upgrade. It would greatly improve maintainability of the code.
* Shiro support: support for Shiro authentication is only partial, as it's limited to basic authentication. I've already commented on this in the following thread https://groups.google.com/forum/#!topic/spark-jobserver/hRgieVUrkQ8.I'd be happy to help there if needed.

**Comment by marcoboi** [comment]
Following up on my previous post, there is one more outstanding issue that would be most convenient addressing at this stage.
* Currently, the maximum size of a file upload is limited by the amount of memory available to the application.
When uploading a file through the "data" API, Job Server will read the file from the HTTP request into a byte array, before dumping it to disk. As the file is loaded into memory, this limits the maximum file upload size to the available memory.
This limitation could be easily removed by writing the stream of the uploaded file to disk without loading it to memory first.
Currently, design constraints makes it difficult to make such a change, as I explained here: https://groups.google.com/forum/#!topic/spark-jobserver/ZU5ac4AdXD0

**Comment by velvia** [comment]
@marcoboi, thanks, agreed with your points.  Some of the points had been tackled in previous PRs but not fully incorporated or gotten lost.  Definitely time to clean all of it up

**Comment by Majea** [comment]
That's a whole lot of points there, @marcoboi. I'm not sure how and when we can help and contribute. 
@htnd187, should we wait for the end of the akka-http refactoring and then contribute with PRs if some of those points are not covered? Should be try to help directly with PRs to akka-http-expirimental branch? Any other idea?

**Comment by hntd187** [comment]
@Majea I'm not sure what the right answer is there because I don't know when this will ultimately get finished, but at the same time if you have a pressing need for some new functionality I don't want to hold things up. Because this is a pretty large, high impact change I want to take the time to make sure we do it right, I don't think there is any reason to rush this along. 

As for @marcoboi 's suggestions those are all fantastic ones and I think they are ones we can do. I think we can also squash the max file upload bug too. I'm almost done hashing out a first pass on the jobs route just to finish a set of routes to setup the tests and functionality and all of that. People can also comment on the structure I got so far and then we can move from there.

**Comment by marcoboi** [comment]
@Majea , @hntd187 , @velvia ,
I believe some of those issues imply a large refactoring, so we should try address them here. Specifically, I'm referring to points 1, 4 and 6. 
Points 2, 3 and 5 could be addressed in the main branch and successively ported here.
Let me know what you guys think.

**Comment by velvia** [comment]
Sorry @hntd187 not to add to the list but we should try to take care of #887 as well, a easy one to do  :)
